app/services/pages/column.rb

```diff
@@ -0,0 +1,18 @@
+module Pages
+  class Column
+
+    attr_reader :widgets
+
+    def initialize
+      @widgets = []
+    end
+
+    def find_widget(widget_name)
+      @widgets.select { |w| w.name == widget_name.to_sym }.first
+    end
+
+    def <<(widget)
+      @widgets << widget
+    end
+  end
+end
```

Comment:
In both cases (page or tab) the extension point is well-defined, so there can be simple assumption that every single extension point has it's own CSS row. Meaning there are 12 columns available for consumption. Plugins can simple drop elements in it, Bootstrap framework has predefined behavior how to put elements on the page. There can be conflicts of course, but as long as plugins drop elements of decent size (let's say 6) we are fine.

Do you have some working examples of plugins that add:

- one widget onto a page
- four widgets (2x2 grid) onto a page
- the same for tab

Just pastebin me working `extend_page` block.

In short - I think building a framework on top of Bootstrap Grid is not necessary and too complicated. Simple ERB/HTML is enough, the important addition here are the extension points.